### PR TITLE
[SVS-5] Additional files for parametrized SonarQube rules are retrieved from the SonarQube quality profile

### PR DESCRIPTION
**Pull request by mjcheetham**
Added an extra step in the BindingWorkflow to unpack the additional files (if any) that are downloaded with the quality profile export XML. Additional files are unpacked under the SonarQube folder, alongside the solution level rule set file.
Also renamed the IBindingWorkflow interface to IBindingWorkflowExecutor in order to better reflect what it actually does (similar to the IConnectionWorkflowExec..). With the previous name, it was strange to have BindCommand implement IBindingWorkflow, but BindingWorkflow NOT implement IBindingWorkflow.

### CODE REVIEW

File: src/Integration.UnitTests/Binding/BindCommandTests.cs [hunk 1]
```diff
@@ -362,12 +362,12 @@ private BindCommand PrepareCommpandForExecution(ConnectionInformation connection
             return testSubject;
         }
 
-        private class TestBindingWorkflow : IBindingWorkflow
+        private class TestBindingWorkflow : IBindingWorkflowExecutor
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Binding/BindCommandTests.cs]: Need to change the test implementation i.e. ConfigurableBindingWorkflow

File: src/Integration/Binding/BindingWorkflow.cs [hunk 2]
```diff
@@ -172,7 +181,7 @@ private ProgressStepDefinition[] CreateWorkflowSteps(IProgressController control
             }
         }
 
-        internal /*for testing purposes*/ void DownloadRuleSet(IProgressController controller, CancellationToken cancellationToken, IProgressStepExecutionEvents notificationEvents, IEnumerable<string> languages)
+        internal /*for testing purposes*/ void DownloadQualityProfile(IProgressController controller, CancellationToken cancellationToken, IProgressStepExecutionEvents notificationEvents, IEnumerable<string> languages)
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Need also inject the additional items into an itemgroup for each project file.

File: src/Integration/Binding/BindingWorkflow.cs [hunk 3]
```diff
@@ -320,10 +331,45 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
 
+        internal /*for testing purposes*/ void UnpackAdditionalFiles(IProgressController controller, CancellationToken token, IProgressStepExecutionEvents notificationEvents)
+        {
+            if (!this.AdditionalFiles.Any())
+            {
+                return; // Nothing to unpack
+            }
+
+            Solution2 solution = this.projectSystemHelper.GetCurrentActiveSolution();
+            if (solution == null)
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: An assert will be enough IMO. Otherwise you have a bug and a missing test case (need to return a from this if block, otherwise you will get a null ref exception).

File: src/Integration/Binding/BindingWorkflow.cs [hunk 4]
```diff
@@ -320,10 +331,45 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
+            {
+                VsShellUtils.WriteToGeneralOutputPane(this.owner.ServiceProvider, Strings.FailedToUnpackAdditionalFiles);
+                bool aborted = controller.TryAbort();
+                Debug.Assert(aborted || token.IsCancellationRequested, "Failed to abort the workflow");
+            }
+
+            string solutionRoot = Path.GetDirectoryName(solution.FullName);
+            string root = this.solutionRuleSetWriter.GetOrCreateRuleSetDirectory(PathHelper.ForceDirectoryEnding(solutionRoot));
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: PathHelper.ForceDirectoryEnding(solutionRoot) - strange that we need to do it... But the looks of it, is shouldn't [using Path.Combine].
If GetOrCreate... needs it to end with '\' then it needs to call the helper...

File: src/Integration/Resources/Strings.resx [hunk 5]
```diff
@@ -405,4 +405,12 @@ Localization note: '[Don't warn me again]()' is a special markup used to transla
     <value>SonarQube project binding failed: The server does not have the minimum required version ({0}) of the C# plugin installed.</value>
     <comment>Error message indicating that the server does not have the required version of the C# plugin installed. {0} minimum required version</comment>
   </data>
+  <data name="FailedToUnpackAdditionalFiles" xml:space="preserve">
+    <value>Failed to unpack additional files required for analyzers</value>
```
Comment (eyalantzMS) [inline src/Integration/Resources/Strings.resx]: 'required for analyzers': 'specified by the quality profile' might be better and more contextual in terms of the information the user gets from our extension in the output.

File: src/Integration/Binding/BindingWorkflow.cs [hunk 6]
```diff
@@ -195,6 +204,8 @@ private ProgressStepDefinition[] CreateWorkflowSteps(IProgressController control
 
                 this.NuGetPackages.AddRange(export.Deployment.NuGetPackages);
 
+                this.AdditionalFiles.AddRange(export.Configuration.AdditionalFiles);
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: 'AdditionalFiles' can it be null? Same question for NuGetPacakges.

Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: No to both. The XmlSerializer will ensure an empty List<T> is created, even if no T nodes are in the serialised data.

File: src/Integration/ProjectSystemHelper.cs [hunk 7]
```diff
@@ -188,5 +194,10 @@ private static bool IsProjectKind(Project project, string projectKindGuidString)
         {
             return StringComparer.OrdinalIgnoreCase.Equals(projectKindGuidString, project.Kind);
         }
+
+        public Microsoft.Build.Evaluation.Project GetEquivalentMSBuildProject(Project project)
+        {
+            return Microsoft.Build.Evaluation.ProjectCollection.GlobalProjectCollection.LoadedProjects.FirstOrDefault(p => p.FullPath == project.FullName);
```
Comment (mjcheetham) [inline src/Integration/ProjectSystemHelper.cs]: Need to use the MSBuild Project type to add an ProjectItem of a specified type.

File: src/Integration/Binding/BindingWorkflow.cs (hunk 2 above)
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: See new commit

File: src/Integration/Binding/BindingWorkflow.cs (hunk 3 above)
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: Replace with assert and return.

File: src/Integration/Resources/Strings.resx (hunk 5 above)
Comment (mjcheetham) [inline src/Integration/Resources/Strings.resx]: Done.

File: src/Integration.UnitTests/Binding/BindCommandTests.cs (hunk 1 above)
Comment (mjcheetham) [inline src/Integration.UnitTests/Binding/BindCommandTests.cs]: Already done in master (I believe, by yourself)

File: src/Integration.UnitTests/Binding/BindingWorkflowTests.cs [hunk 8]
```diff
@@ -362,6 +444,43 @@ private void ConfigureProfileExport(BindingWorkflow testSubject, RoslynExportPro
...
+            }
+        }
+
+        private static void VerifyAdditionalFileAddedToProject(Project vsProject, MSBuild.Project msbuildProject, string additionalFilePath)
+        {
+            string relativePath = PathHelper.CalculateRelativePath(vsProject.FullName, additionalFilePath);
+
+            var additionalFileItems = msbuildProject.Items.Where(x => x.ItemType == Constants.AdditionalFilePropertyKey);
+
+            Assert.IsTrue(additionalFileItems.Select(x => x.UnevaluatedInclude).Contains(relativePath), "AdditionalFile is missing");
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Binding/BindingWorkflowTests.cs]: Probably worth including in the assert message the relative file that was missing

File: src/Integration/Binding/BindingWorkflow.cs [hunk 9]
```diff
@@ -320,10 +339,84 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
+                foreach (var additionalFilePath in this.AdditionalFilePaths)
+                {
+                    this.projectSystemHelper.AddFileToProject(solutionItemsProject, additionalFilePath);
+                }
+            }
+
+            // Inject into projects
+            foreach (var project in this.projectSystemHelper.GetSolutionManagedProjects())
+            {
+                string projectRoot = PathHelper.ForceDirectoryEnding(Path.GetDirectoryName(project.FullName));
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Shouldn't CalculateRelativePath force directory ending instead of sprinkling this in the calling code?

File: src/Integration/Binding/BindingWorkflow.cs [hunk 10]
```diff
@@ -320,10 +339,84 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
+
+                MSBuild.Project msBuildProject = this.projectSystemHelper.GetEquivalentMSBuildProject(project);
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Assert that msBuildProject is not null

File: src/Integration/Binding/BindingWorkflow.cs (hunk 9 above)
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: Assuming that CalculateRelativePath is only going to be used to calculate the relative path from a directory, and not a file with the same name.
Currently CalcRelPath will also calculate paths relative to files as well as directories.

File: src/Integration/Binding/BindingWorkflow.cs [hunk 11]
```diff
@@ -320,10 +339,84 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
+
+                var existingItems = msBuildProject.Items.Where(x => x.ItemType == Constants.AdditionalFilePropertyKey);
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Might be better to have a hash set of EvalatedIncludes (will be faster and more straight forward), also the hash set needs to be string ordinal ignore case (since it's files) which the current code is not handling. Also add a test case for existing additional files with different casing.

File: src/Integration/Binding/BindingWorkflow.cs [hunk 12]
```diff
@@ -320,10 +339,84 @@ private void FinishBindingOnUIThread(IProgressController controller, IProgressSt
...
+
+                foreach (var additionalFilePath in this.AdditionalFilePaths)
+                {
+                    var existingItem = existingItems.FirstOrDefault(x => x.EvaluatedInclude == additionalFilePath);
+
+                    if (existingItem == null)
+                    {
+                        string relativePath = PathHelper.CalculateRelativePath(projectRoot, additionalFilePath);
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Worth commenting how is that you search for the additional file but when adding you add a relative path (msdn doesn't have enough documentation it seems, and you figured out what is expected...)

File: src/Integration/ProjectSystemHelper.cs (hunk 7 above)
Comment (eyalantzMS) [inline src/Integration/ProjectSystemHelper.cs]: Ordinal ignore case...

**Comment by eyalantzMS** [comment]
A few issues that needed to be addressed, otherwise LGTM.

File: src/Integration/Binding/BindingWorkflow.cs (hunk 9 above)
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: (didn't check) worth documenting ...

File: src/Integration.UnitTests/Binding/BindingWorkflowTests.cs (hunk 8 above)
Comment (mjcheetham) [inline src/Integration.UnitTests/Binding/BindingWorkflowTests.cs]: Done

File: src/Integration/ProjectSystemHelper.cs (hunk 7 above)
Comment (mjcheetham) [inline src/Integration/ProjectSystemHelper.cs]: Done

File: src/Integration/Binding/BindingWorkflow.cs (hunk 10 above)
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: Done

File: src/Integration/Binding/BindingWorkflow.cs (hunk 11 above)
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: Done.

**Comment by Xdjkstylesx** [comment]
> 😉👍
